File: src/detect_readthrough_exons.pl
#!/usr/bin/env perl
$protein_gff=$ARGV[0];

open(FILE,$protein_gff);
while($line=<FILE>){
  chomp($line);
  @F=split(/\t/,$line);
  if($F[2] eq "gene"){
    @f=split(/;/,$F[8]);
    $pbeg{substr($f[0],3)}=$F[3];
    $pend{substr($f[0],3)}=$F[4];
  }
}

my $protid="";
my $transcript="";
while($line=<STDIN>){
  chomp($line);
  @F=split(/\t/,$line);
  if($F[2] eq "transcript"){
    $flag=0;
    if($F[8]=~/^transcript_id "(\S+)"; gene_id (.+); cmp_ref "(\S+)"; class_code "(k|=|j|c|m|n|o|i|y)"/){
      $protid=$3;
      $tid=$1;
      @tr=split(/\./,$1);
      $transcript=join(".",@tr[0..($#tr-1)])." ".$tr[-1];
      $flag=1;
      #print "DEBUG $transcript $F[3] $F[4] $pbeg{$protid} $pend{$protid} $flag\n";
    }
  }elsif($F[2] eq "exon" && $flag){
    #print "DEBUG exon $transcript $F[3] $F[4] $F[6] $pbeg{$protid} $pend{$protid}\n";
    print "$transcript $F[3] $F[4] $F[6] $pbeg{$protid} $pend{$protid}\n" if((($tid =~ /5p$/ && $pend{$protid}>=$F[3])||($tid =~ /3p$/ && $pbeg{$protid}<=$F[4])) && $F[6] eq "+");
    print "$transcript $F[3] $F[4] $F[6] $pbeg{$protid} $pend{$protid}\n" if((($tid =~ /5p$/ && $pbeg{$protid}<=$F[4])||($tid =~ /3p$/ && $pend{$protid}>=$F[3])) && $F[6] eq "-");
  }
}

File: src/detect_readthrough_exons.py
#!/usr/bin/env python
#this code detects which exons in UTRs are possible readthroughs
import re
import sys

pbeg={}
pend={}
protein_gff=sys.argv[1]
with open(protein_gff,"r") as f:
    for line in f:
        if line[0] == "#": continue
        gff_fields=line.strip().split("\t")
        if len(gff_fields) !=9: continue 
        if gff_fields[2] == "gene":
            gff_attr=gff_fields[8].split(";")
            #print(gff_attr[0][3:])
            pbeg.update({gff_attr[0][3:]:int(gff_fields[3])})
            pend.update({gff_attr[0][3:]:int(gff_fields[4])})


for line in sys.stdin:
    if line[0] == "#": continue
    gff_fields=line.strip().split("\t")
    if len(gff_fields) !=9: continue
    gff_fields[3]=int(gff_fields[3])
    gff_fields[4]=int(gff_fields[4])
    if gff_fields[2] == "transcript":
        flag=False
        #print(gff_fields[8])
        x=re.search(r"class_code \"\S+\"",gff_fields[8])
        class_code=x.group()[12:-1]
        if re.search(class_code,"k=jcmnoiy") == None: continue
        x=re.search(r"transcript_id \"\S+\"",gff_fields[8])
        tid=x.group()[15:-1]
        x=re.search(r"cmp_ref \"\S+\"",gff_fields[8])
        protid=x.group()[9:-1]
        #print(tid+" "+protid+" "+class_code)
        tr=tid.split(".")
        transcript=".".join(tr[0:-1])+" "+tr[-1]
        #print(transcript)
        if ((tr[-1]=="5p" and pend[protid]<=gff_fields[4]) or (tr[-1]=="3p" and pbeg[protid]>=gff_fields[3])) and gff_fields[6] =="+": flag=True
        if ((tr[-1]=="5p" and pbeg[protid]>=gff_fields[3]) or (tr[-1]=="3p" and pend[protid]<=gff_fields[4])) and gff_fields[6] =="-": flag=True
    elif gff_fields[2] == "exon" and flag:
        if (((tr[-1]== "5p" and pend[protid]>gff_fields[3]) or (tr[-1] == "3p" and pbeg[protid]<gff_fields[4])) and gff_fields[6] == "+") or (((tr[-1]== "5p" and pbeg[protid]<gff_fields[4]) or (tr[-1] == "3p" and pend[protid]>gff_fields[3])) and gff_fields[6] == "-"):
            print(transcript+" "+str(gff_fields[3])+" "+str(gff_fields[4])+" "+gff_fields[6]+" "+str(pbeg[protid])+" "+str(pend[protid]))


